--- 
title: "JSEQ® REPORT"
date: "`r Sys.Date()`"
output: html_document
---


```{r include=FALSE}

library(Seurat)
library(tidyverse)
library(patchwork)
library(gridExtra)
library(grid)
library(knitr)
library(viridis)


```

\

### **The report was generated in JSEQ® single-cell analysis pipeline**
##### **For more information go to https://github.com/jkubis96/JSEQ_scRNAseq and read JSEQ_manual.pdf**

\

```{r message=FALSE, warning=FALSE, include=FALSE}
args <- commandArgs()

path <- args[6]
markers <-args[7]
species <- args[8]
seurat_umi <- file.path(path,'sc_data/')
OUTPUT <- file.path(path, 'results')
path_results <- file.path(path, 'results')
project_name <- args[9]
data <- args[10]
estimated_cells <- args[11]
tissue <- args[12]
cell_development_status <- args[13]
functions <- file.path(getwd(), 'functions.R')
source(functions)

path_tmp <- file.path(path, 'results')

UMI <- readRDS(file = file.path(path_results, "Results.rds"))
cells_number <- length(UMI@active.ident)

metrics <- file.path(path_tmp,'scRNAmetrics.txt')
chceck_fq <- grepl("_fq", metrics)
```

### **The cell content analysis**

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", dpi=600, retina=1,  results='asis'}

if (chceck_fq == TRUE){
  
cat('**The conntent of mapped reads of the detected barcodes**')
  
mydata <- read.csv(file = metrics, header = T,
                   stringsAsFactors = F, skip = 6, sep = "\t")
mydata <- mydata[order(mydata$PF_ALIGNED_BASES, decreasing = T), ]
mydata_pct <- mydata[, c("READ_GROUP",
                         "PCT_INTERGENIC_BASES",
                         "PCT_UTR_BASES",
                         "PCT_RIBOSOMAL_BASES",
                         "PCT_INTRONIC_BASES",
                         "PCT_CODING_BASES")
                     ]
colnames(mydata_pct) = c('Cell Barcode', 'Intergenic', 'UTR', 'Ribosomial', 'Intronic', 'Coding')

mydata <- mydata[, c("READ_GROUP",
                     "INTERGENIC_BASES",
                     "UTR_BASES",
                     "RIBOSOMAL_BASES",
                     "INTRONIC_BASES",
                     "CODING_BASES")
                 ]
colnames(mydata) = c('Cell Barcode', 'Intergenic', 'UTR', 'Ribosomial', 'Intronic', 'Coding')

# converting into long format for ploting
mydata_long <- mydata %>% gather("Read Overlap", count, -"Cell Barcode")

# Keep the original order of the barcodes using factor and levels.
mydata_long$`Cell Barcode` <- factor(mydata_long$`Cell Barcode`,
                                 levels = factor(unique(mydata_long$`Cell Barcode`)))
mydata_long$`Read Overlap` <- factor(mydata_long$`Read Overlap`,
                                   levels = unique(mydata_long$`Read Overlap`))

p1 <- ggplot(mydata_long, aes(x = `Cell Barcode`, y = count, fill = `Read Overlap`)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0), legend.position = "none")
p1 <- p1 + labs(x = "Barcodes", y = "Bases")
p1 <- p1 + theme(axis.title.x = element_blank(),
                 axis.text.x = element_blank(),
                 axis.ticks.x = element_blank())
p1 <- p1 + scale_y_continuous(labels = scales::scientific)
p1 <- p1 + scale_fill_viridis(discrete = TRUE, option = "viridis")


mydata_long_pct <- mydata_pct %>% gather("Read Overlap", fraction, -"Cell Barcode")
# Keep the original order of the barcodes using factor and levels.
mydata_long_pct$`Cell Barcode` <- factor(mydata_long_pct$`Cell Barcode`,
                                     levels = factor(unique(mydata_long_pct$`Cell Barcode`)))
mydata_long_pct$`Read Overlap` <- factor(mydata_long_pct$`Read Overlap`,
                                       levels = unique(mydata_long_pct$`Read Overlap`))

p2 <- ggplot(mydata_long_pct, aes(x = `Cell Barcode`, y = fraction, fill = `Read Overlap`)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 0, size=8, vjust = 0.05), legend.position = "bottom") +
  labs(x = "Barcodes", y = "%Bases") +
  scale_y_continuous(labels = scales::percent) + scale_fill_viridis(discrete = TRUE, option = "viridis")


p1

}
```


```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", dpi=600, retina=1}

if (chceck_fq == TRUE) {
p2
}

```

\

```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
if (chceck_fq == TRUE){
cat('**The knee plot for computed discovery the right cells and user defined number of cells based [UMI/BARCODE]**')
}
```

```{r pressure, echo=FALSE, out.width = '70%', fig.align="center"}

if (chceck_fq == TRUE){
  
knitr::include_graphics(file.path(path_results,'expect_whitelist_cell_barcode_counts.png'), dpi = 600)
  
}
``` 


```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

if (chceck_fq == TRUE){
  
cat('*The above graph shows only trend and quality at the beginning of single cell analysis, however, it is not the final result for obtained number of cells')

}
```

\

##### **The number of the genes and counts per the cell**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
knitr::include_graphics(file.path(path_results,'counts~genes.svg'), dpi = 600)
```

\

##### **The ratio of the number of genes to the number of counts**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'counts~genes_QC.svg'), dpi = 600)

```


\

##### **The percent of the ribosomal and mitochondrial genes [%]**


\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'Ribo~Mito.svg'), dpi = 600)

```


\

### **The quality control of the cells**

\

##### **Number of the counts**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'DropletQC.svg'), dpi = 600)

```

\

##### **The content of the mitochondrial genes**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'MitoQC.svg'), dpi = 600)

```


\

##### **The content of the ribosomal genes**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'RiboQC.svg'), dpi = 600)

```

\

##### **Number of the cells at the different steps of the analysis**

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'Cells.svg'), dpi = 600)


```

\
\

##### **Top of the high variable genes in the data set**

\

```{r echo=FALSE, message=FALSE, warning=FALSE}

top20 <- head(VariableFeatures(UMI), 20)
top20

```

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", dpi=600, retina=1}

plot1 <- VariableFeaturePlot(UMI)
LabelPoints(plot = plot1, points = top20, repel = TRUE)


```

\

```{r message=FALSE, warning=FALSE, include=FALSE}

Elbow <- ElbowPlot(UMI, ndims = 50)

dims <- as.data.frame(Elbow$data$stdev)

#select the most variable reduction

{
  dim <- 1
  score <- c()
  element <- 0
  for (i in dims$`Elbow$data$stdev`) {
    element <- element + 1
    if (i-i*0.01 > dims$`Elbow$data$stdev`[element+1] & element < 50 | i-i*0.02 > dims$`Elbow$data$stdev`[element+2] & element < 49 | i-i*0.02 > dims$`Elbow$data$stdev`[element+3] & element < 48 | i-i*0.02 > dims$`Elbow$data$stdev`[element+4] & element < 47) {
      dim <- dim + 1
    } else 
      break
  }
  dim <- as.numeric(dim)
}


```


\

### **Analysis parameter selection**

\

##### **The cut off for the number of the principal components**

\

##### **ElbowPlot**

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", dpi=600, retina=1}
Elbow + geom_vline(xintercept = dim, color = 'red')
```

\

##### **JackStrawPlot for `r dim` PCs**

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
knitr::include_graphics(file.path(path_results,'JackStrawPlot.svg'), dpi = 600)
```

\

### **The results for the cells grouping**
##### **The UMAP processing and presentation of the cell clusters**

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
knitr::include_graphics(file.path(path_results,'UMAP.svg'), dpi = 600)
```

##### **The cluster selected markers**

```{r echo=FALSE, message=FALSE, warning=FALSE, align="center"}
markers <- read.csv(file = file.path(path_results, "subclasses_marker_report.csv"), header = T, sep = ',', row.names = 1)
knitr::kable(markers, align = "l")
```

\
\

### **The cluster classification and naming via the markers**

\

##### **The cell subclasses**

\

##### **The PC plot**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'PCA_DimPlot_class.svg'), dpi = 600)

```

\

##### **The UMAP plot**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'UMAP_with_DE_gene_class.svg'), dpi = 600)

```

\

### **The presentation of the subtypes obtained using the CSSG algorithm**

\

##### **The quality control of the cell subtypes**

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}

knitr::include_graphics(file.path(path_results,'cells_type_threshold.svg'), dpi = 600)

```

\

##### **The PC plot**

\


```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", dpi=600, retina=1}

knitr::include_graphics(file.path(path_results,'PCA_DimPlot_subtypes.svg'), dpi = 600)

```

\

##### **The UMAP plot**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", dpi=600, retina=1}

knitr::include_graphics(file.path(path_results,'UMAP_with_DE_gene_subtypes.svg'), dpi = 600)

```

\

##### **The HDMAP plot**

\

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", dpi=600, retina=1}

knitr::include_graphics(file.path(path_results,'HDMAP_subtypes.svg'), dpi = 600)


```

\

### **The subtypes selected markers**

```{r echo=FALSE, message=FALSE, warning=FALSE, align="center"}
markers <- read.csv(file = file.path(path_results, "subtypes_marker_report.csv"), header = T, sep = ',', row.names = 1)
markers <- markers[,colnames(markers) %in% c('p_val','avg_logFC', 'cluster', 'gene')]
knitr::kable(markers, align = "l")
```

\
\

##### **The heatmap of the cell subtypes with specific markers**
###### **The expression normalization by log(CPM+1)**

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align= "center"}

knitr::include_graphics(file.path(path_results,'pheatmap_cells_populations.svg'), dpi = 600)

```

\

### **The cell subtypes heterogeneity - number of genes vs mean expression**

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.align= "center"}

knitr::include_graphics(file.path(path_results,'cells_heterogenity.svg'), dpi = 600)

```


\
\

## **Disclaimer**

```{r echo=FALSE, message=FALSE, warning=FALSE, results='asis'}

cat('The analysis was performed through the non-profit and open-source pipeline JSEQ®. Pipeline belongs to the part of the JBioSystem project in which we create complex software for bioinformatics analysis. Remember in silico analysis can not be taken for granted and must be validated in vitro or in vivo. For more information go to GitHub or contact us. The JSEQ® was created in the Department of Molecular Neurobiology, Institute of Bioorganic Chemistry, Polish Academy of Sciences')

```
\
\